test/unit/services/finance/billing_service_test.rb

```diff
@@ -33,33 +37,45 @@ class Finance::BillingServiceTest < ActionDispatch::IntegrationTest
     end
   end
 
-  test 'creates a lock' do
+  test 'lock remains if execution fails' do
     now = Time.utc(2018, 1, 16, 8)
-    lock = BillingLock.create(account_id: @provider.id)
-    BillingLock.expects(:create).returns(lock)
-    assert Finance::BillingService.call!(@provider.id, now: now)
+    Finance::BillingStrategy.expects(:daily).raises RuntimeError, 'random failure'
+    assert_raises RuntimeError do
+      Finance::BillingService.call!(@provider.id, now: now)
+    end
+    Finance::BillingService.any_instance.expects(:report_error).with { |error| error.is_a? Finance::BillingService::LockBillingError }
+    Finance::BillingService.call!(@provider.id, now: now)
   end
 
+  # WARNING: flakiness here means a bug
   test 'lock prevents multiple jobs from running' do
-    now = Time.utc(2018, 1, 16, 8)
-    BillingLock.create!(account_id: @provider.id)
-    refute Finance::BillingService.call!(@provider.id, now: now)
-  end
+    finished = false
+    lock_thread = within_async_thread do
+      Finance::BillingService.new(@provider.id).send(:with_lock) do
+        sleep 0.1 until finished
+      end
+    end
 
-  test 'lock is released after execution' do
-    now = Time.utc(2018, 1, 16, 8)
-    assert_no_difference BillingLock.method(:count) do
-      Finance::BillingService.call!(@provider.id, now: now)
+    # this thread will prevent a test case deadlock by stopping the lock thread in
+    # case concurrent lock sits waiting for lock to release instead of failing quick
+    safety_thread = Thread.new do
+      Thread.current.report_on_exception = false
+      next if lock_thread.join(5)
+
+      finished = true
+      raise "billing locking causes concurrent billing jobs to wait but does not cancel them"
     end
-  end
 
-  test 'lock is released if execution fails' do
-    now = Time.utc(2018, 1, 16, 8)
-    Finance::BillingStrategy.expects(:daily).raises RuntimeError, 'random failure'
-    BillingLock.expects(:delete).with(@provider.id).returns(true)
-    assert_raises RuntimeError do
-      Finance::BillingService.call!(@provider.id, now: now)
+    sleep 0.1 # make sure first thread has time to acquire the lock
+
+    assert_raise(Finance::BillingService::LockBillingError) do
+      Finance::BillingService.new(@provider.id).send(:with_lock) do
+        safety_thread.join(0.001) # this will raise if thread raised so we know what actually happened
```

Comment:
```suggestion
        # we shouldn't reach here because +lock_thread+ holds the lock, so +LockBillingError+ should have been raised and this block never been executed
        safety_thread.join(0.001) # this will raise if +lock_thread+ raised so we know what actually happened
```

Reply:
I hope this is clearer. What I missed is an explanation to not have to spend time figuring out how the test works the next time.